### PR TITLE
feat: allow upgrades in maintenance mode (only over SideroLink)

### LINKED ISSUE #6224: allow upgrades in maintenance mode (only over SideroLink)
Implement Upgrade API in maintenance mode (secured by only coming from the SideroLink addresses).

Allows to perform a node upgrade (without machine config) to a new version. 

This is a limited interim solution to allow upgrading nodes in maintenance mode.

There are two cases for maintenance mode:

1. Booted from the system disk (Talos is installed) - upgrade is possible.
2. Booted from an ISO, PXE-booted, etc. - upgrade is not possible.

We want to handle case (1).

## Implementation Details

Implement `Upgrade` API (limited to only basic upgrades) in the maintenance API service.

Do a pre-check that Talos is installed (if not, fail the API call).

Don't pipe the machine config (as it's missing) to the `installer`.

Should be backportable to 1.2.x.

API should be only available over SideroLink.

Automated tests - not clear if it's possible, we need extensive manual testing.

### PR DESCRIPTION
**Pull request by smira**
This implements a simple way to upgrade Talos node running in maintenance mode (only if Talos is installed, i.e. if `STATE` and `EPHEMERAL` partitions are wiped).

Upgrade is only available over SideroLink for security reasons.

Upgrade in maintenance mode doesn't support any options, and it works without machine configuration, so proxy environment variables are not available, registry mirrors can't be used, and extensions are not installed.

Fixes #6224

Signed-off-by: Andrey Smirnov <andrey.smirnov@talos-systems.com>

### CODE REVIEW

File: internal/app/machined/internal/server/v1alpha1/v1alpha1_server.go [hunk 1]
```diff
@@ -125,7 +118,7 @@ func (s *Server) checkSupported(feature runtime.ModeCapability) error {
 	mode := s.Controller.Runtime().State().Platform().Mode()
 
 	if !mode.Supports(feature) {
-		return fmt.Errorf("method is not supported in %s mode", mode.String())
+		return status.Errorf(codes.FailedPrecondition, "method is not supported in %s mode", mode.String())
```
Comment (frezbo) [inline internal/app/machined/internal/server/v1alpha1/v1alpha1_server.go:121]: :+1:

File: internal/app/machined/main.go [hunk 2]
```diff
@@ -227,25 +227,37 @@ func run() error {
 		log.Printf("controller runtime finished")
 	}()
 
+	// Inject controller into maintenance service.
+	maintenance.InjectController(c)
+
+	initializeCanceled := false
```
[the remark this answers is not in the document]
Comment (smira) [reply internal/app/machined/main.go:233]: nope, see below - error is `context.Canceled`

[the remark this answers is not in the document]
Comment (frezbo) [reply internal/app/machined/main.go:233]: arggg UK english, lol :sweat_smile:

**Comment by smira** [comment]
full logs:

<details>

```
[    4.754263] [talos] next renewal in 11h59m59.606973051s
[    4.754953] [talos] task loadConfig (1/1): this machine is reachable at:
[    4.755795] [talos] task loadConfig (1/1):  fdae:41e4:649b:9303:20e4:1dd7:889b:9136
[    4.756745] [talos] task loadConfig (1/1): server certificate fingerprint:
[    4.757614] [talos] task loadConfig (1/1):  wtjdIZjaKpTF22k9fXUeleTT+dUNwqV9GFiTIR1Z+9o=
[    4.758615] [talos] task loadConfig (1/1): 
[    4.759167] [talos] task loadConfig (1/1): upload configuration using talosctl:
[    4.760081] [talos] task loadConfig (1/1):  talosctl apply-config --insecure --nodes fdae:41e4:649b:9303:20e4:1dd7:889b:9136 --file <config.yaml>
[    4.761699] [talos] task loadConfig (1/1): or apply configuration using talosctl interactive installer:
[    4.762876] [talos] task loadConfig (1/1):  talosctl apply-config --insecure --nodes fdae:41e4:649b:9303:20e4:1dd7:889b:9136 --mode=interactive
[    4.764459] [talos] task loadConfig (1/1): optionally with node fingerprint check:
[    4.765408] [talos] task loadConfig (1/1):  talosctl apply-config --insecure --nodes fdae:41e4:649b:9303:20e4:1dd7:889b:9136 --cert-fingerprint 'wtjdIZjaKpTF22k9fXUeleTT+dUNwqV9GFiTIR1Z+9o=' --file <config.yaml>
[    5.069027] [talos] adjusting time (slew) by 78.324979ms via 188.93.90.21, state TIME_OK, status STA_PLL | STA_NANO {"component": "controller-runtime", "controller": "time.SyncController"}
[   11.588400] OK [/machine.MachineService/Version] 10.155µs unary Success (:authority=[fdae:41e4:649b:9303:20e4:1dd7:889b:9136]:50000;content-type=application/grpc;runtime=Talos;user-agent=grpc-go/1.49.0)
[   11.593431] [talos] upgrade request received: "ghcr.io/smira/installer:maint-upgrade"
[   11.593959] OK [/machine.MachineService/Upgrade] 545.506µs unary Success (:authority=[fdae:41e4:649b:9303:20e4:1dd7:889b:9136]:50000;content-type=application/grpc;runtime=Talos;user-agent=grpc-go/1.49.0)
[   11.595431] [talos] task loadConfig (1/1): failed: failed to receive config via maintenance service: maintenance service failed: context canceled
[   11.597365] [talos] phase config (6/7): failed
[   11.598063] [talos] initialize sequence: failed
[   11.598737] [talos] maintenanceUpgrade sequence: 8 phase(s)
[   11.599655] [talos] phase containerd (1/8): 1 tasks(s)
[   11.600389] [talos] task startContainerd (1/1): starting
[   11.601034] [talos] service[containerd](Preparing): Running pre state
[   11.601775] [talos] service[containerd](Preparing): Creating service runner
[   12.574640] [talos] service[containerd](Running): Process Process(["/bin/containerd" "--address" "/system/run/containerd/containerd.sock" "--state" "/system/run/containerd" "--root" "/system/var/lib/containerd"]) started with PID 713
[   12.623927] [talos] service[containerd](Running): Health check successful
[   12.624538] [talos] task startContainerd (1/1): done, 1.02763208s
[   12.625048] [talos] phase containerd (1/8): done, 1.028881163s
[   12.625551] [talos] phase verifyDisk (2/8): 1 tasks(s)
[   12.625990] [talos] task verifyDiskAvailability (1/1): starting
[   12.626521] [talos] task verifyDiskAvailability (1/1): done, 532.288µs
[   12.627071] [talos] phase verifyDisk (2/8): done, 1.525777ms
[   12.627540] [talos] phase upgrade (3/8): 1 tasks(s)
[   12.628169] [talos] task upgrade (1/1): starting
[   12.799108] [talos] task upgrade (1/1): performing upgrade via "ghcr.io/smira/installer:maint-upgrade"
[   12.802489] [talos] pulling "ghcr.io/smira/installer:maint-upgrade"
[   40.035774] 2022/09/22 14:32:43 running Talos installer v1.2.0-alpha.2-116-g578a18c98-dirty
[   40.036485] 2022/09/22 14:32:43 machine configuration missing, skipping validation
[   40.038365] XFS (vda3): Mounting V5 Filesystem
[   40.046897] XFS (vda3): Ending clean mount
[   40.047878] XFS (vda3): Unmounting Filesystem
[   40.064345] XFS (vda3): Mounting V5 Filesystem
[   40.071145] XFS (vda3): Ending clean mount
[   43.972001] XFS (vda3): Unmounting Filesystem
[   43.973008] 2022/09/22 14:32:47 preserved contents of "BOOT": 141290360 bytes
[   43.976649] 2022/09/22 14:32:47 preserved contents of "META": 1195 bytes
[   44.268534] 2022/09/22 14:32:47 preserved contents of "STATE": 101942 bytes
[   44.269173] 2022/09/22 14:32:47 resetting partition table on /dev/vda
[   44.271485] 2022/09/22 14:32:47 partitioning /dev/vda - EFI "105 MB"
[   44.271920] 2022/09/22 14:32:47 created /dev/vda1 (EFI) size 204800 blocks
[   44.272370] 2022/09/22 14:32:47 partitioning /dev/vda - BIOS "1.0 MB"
[   44.272847] 2022/09/22 14:32:47 created /dev/vda2 (BIOS) size 2048 blocks
[   44.273310] 2022/09/22 14:32:47 partitioning /dev/vda - BOOT "1.0 GB"
[   44.273737] 2022/09/22 14:32:47 created /dev/vda3 (BOOT) size 2048000 blocks
[   44.274253] 2022/09/22 14:32:47 partitioning /dev/vda - META "1.0 MB"
[   44.274677] 2022/09/22 14:32:47 created /dev/vda4 (META) size 2048 blocks
[   44.275121] 2022/09/22 14:32:47 partitioning /dev/vda - STATE "105 MB"
[   44.275616] 2022/09/22 14:32:47 created /dev/vda5 (STATE) size 204800 blocks
[   44.276098] 2022/09/22 14:32:47 partitioning /dev/vda - EPHEMERAL "0 B"
[   44.276807] 2022/09/22 14:32:47 created /dev/vda6 (EPHEMERAL) size 10117120 blocks
[   44.277630] 2022/09/22 14:32:47 formatting the partition "/dev/vda1" as "vfat" with label "EFI"
[   44.289352] 2022/09/22 14:32:47 zeroing out "/dev/vda2"
[   44.290449] 2022/09/22 14:32:47 formatting the partition "/dev/vda3" as "xfs" with label "BOOT"
[   44.295470] 2022/09/22 14:32:47 zeroing out "/dev/vda4"
[   44.296390] 2022/09/22 14:32:47 zeroing out "/dev/vda5"
[   44.366994] 2022/09/22 14:32:47 zeroing out "/dev/vda6"
[   44.368111] XFS (vda3): Mounting V5 Filesystem
[   44.370623] XFS (vda3): Ending clean mount
[   44.656583] XFS (vda3): Unmounting Filesystem
[   44.658792] 2022/09/22 14:32:47 restored contents of "BOOT"
[   44.660721] 2022/09/22 14:32:47 restored contents of "META"
[   44.766850] 2022/09/22 14:32:47 restored contents of "STATE"
[   44.768031] XFS (vda3): Mounting V5 Filesystem
[   44.772378] XFS (vda3): Ending clean mount
[   44.774254] 2022/09/22 14:32:47 copying /usr/install/amd64/vmlinuz to /boot/B/vmlinuz
[   44.786458] 2022/09/22 14:32:47 copying /usr/install/amd64/initramfs.xz to /boot/B/initramfs.xz
[   44.827446] 2022/09/22 14:32:47 writing /boot/grub/grub.cfg to disk
[   44.828422] 2022/09/22 14:32:47 executing: grub-install --boot-directory=/boot --efi-directory=/boot/EFI --removable /dev/vda
[   44.830076] Installing for x86_64-efi platform.
[   44.943300] Installation finished. No error reported.
[   45.132804] XFS (vda3): Unmounting Filesystem
[   45.135586] 2022/09/22 14:32:48 installation of v1.2.0-alpha.2-116-g578a18c98-dirty complete
[   45.201966] [talos] task upgrade (1/1): upgrade successful
[   45.202569] [talos] task upgrade (1/1): done, 32.607252377s
[   45.203258] [talos] phase upgrade (3/8): done, 32.608572274s
[   45.203832] [talos] phase mountBoot (4/8): 1 tasks(s)
[   45.204363] [talos] task mountBootPartition (1/1): starting
[   45.374932] XFS (vda3): Mounting V5 Filesystem
[   45.380153] XFS (vda3): Ending clean mount
[   45.381084] [talos] task mountBootPartition (1/1): done, 176.54285ms
[   45.381955] [talos] phase mountBoot (4/8): done, 177.948508ms
[   45.382791] [talos] phase kexec (5/8): 1 tasks(s)
[   45.383441] [talos] task kexecPrepare (1/1): starting
[   45.384247] [talos] task kexecPrepare (1/1): done, 803.724µs
[   45.384984] [talos] phase kexec (5/8): done, 2.193527ms
[   45.385658] [talos] phase unmountBoot (6/8): 1 tasks(s)
[   45.386355] [talos] task unmountBootPartition (1/1): starting
[   45.387416] XFS (vda3): Unmounting Filesystem
[   45.388670] [talos] task unmountBootPartition (1/1): done, 2.311564ms
[   45.389563] [talos] phase unmountBoot (6/8): done, 3.901097ms
[   45.390317] [talos] phase stopEverything (7/8): 1 tasks(s)
[   45.391084] [talos] task stopAllServices (1/1): starting
[   45.392015] [talos] service[containerd](Stopping): Sending SIGTERM to Process(["/bin/containerd" "--address" "/system/run/containerd/containerd.sock" "--state" "/system/run/containerd" "--root" "/system/var/lib/containerd"])
[   45.399579] [talos] service[containerd](Finished): Service finished successfully
[   45.400710] [talos] task stopAllServices (1/1): done, 9.614128ms
[   45.401553] [talos] phase stopEverything (7/8): done, 11.221367ms
[   45.402460] [talos] phase reboot (8/8): 1 tasks(s)
[   45.403191] [talos] task reboot (1/1): starting
[   45.504436] [talos] rebooting in 10 seconds
```

</details>

**Review by frezbo — APPROVED** [review]

File: internal/app/machined/pkg/runtime/v1alpha1/v1alpha1_controller.go [hunk 3]
```diff
@@ -370,6 +370,17 @@ func (c *Controller) phases(seq runtime.Sequence, data interface{}) ([]runtime.P
...
 		phases = c.s.StageUpgrade(c.r, in)
+	case runtime.SequenceMaintenanceUpgrade:
+		var (
+			in *machine.UpgradeRequest
+			ok bool
+		)
+
+		if in, ok = data.(*machine.UpgradeRequest); !ok {
+			return nil, runtime.ErrInvalidSequenceData
+		}
```
Comment (DmitriyMV) [inline internal/app/machined/pkg/runtime/v1alpha1/v1alpha1_controller.go:381]:
```suggestion
		in, ok := data.(*machine.UpgradeRequest);
		if !ok {
			return nil, runtime.ErrInvalidSequenceData
		}
```

Minor nit: this looks shorter

File: internal/pkg/install/install.go [hunk 4]
```diff
@@ -205,19 +216,37 @@ func RunInstallerContainer(disk, platform, ref string, cfg config.Provider, opts
...
+		configR = &containerdrunner.StdinCloser{
+			Stdin:  bytes.NewReader(configBytes),
+			Closer: make(chan struct{}),
+		}
+
+		r = configR
 	}
 
-	creator := cio.NewCreator(cio.WithStreams(configR, w, w))
+	creator := cio.NewCreator(cio.WithStreams(r, w, w))
```
Comment (DmitriyMV) [inline internal/pkg/install/install.go:240]: Can't we use `configR` here?

**Review by DmitriyMV — APPROVED** [review]

File: internal/pkg/install/install.go (hunk 4 above)
Comment (smira) [reply internal/pkg/install/install.go:240]: `r` is `io.Reader`, so `r == nil` will be true (if it's `nil`)
`configR` is `*StdinCloser`, so if it's `nil`, then if passed as `io.Reader` it won't compare to `nil` (`r != nil`)

Comment (DmitriyMV) [reply internal/pkg/install/install.go:240]: Maybe then 
```go
	var r interface {
		io.Reader
		WaitAndClose(ctx context.Context, task containerd.Task)
	}
```

Comment (smira) [reply internal/pkg/install/install.go:240]: this should probably work

File: internal/app/machined/pkg/runtime/v1alpha1/v1alpha1_controller.go (hunk 3 above)
Comment (smira) [reply internal/app/machined/pkg/runtime/v1alpha1/v1alpha1_controller.go]: fixed

File: internal/pkg/install/install.go (hunk 4 above)
Comment (smira) [reply internal/pkg/install/install.go:238]: fixed, thanks!

**Pull request merged**